documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,486 @@
+reingestion DAG is reingesting them at the same time the `refresh_flickr_scores`
+task is refreshing their popularity scores. Concurrent updates might happen in
+this order:
+
+| Flickr Reingestion DAG | `refresh_flickr_scores` | Comment                                                                                                                                                               |
+| ---------------------- | ----------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Upsert record A        | Update record B         |                                                                                                                                                                       |
+| Upsert record B        |                         | `refresh_flickr_scores` has a lock on the row containing record B, so the Flickr Reingestion DAG must wait.                                                           |
+|                        | Update record A         | The Flickr Reingestion DAG has a lock on the row containing record A, so `refresh_flickr_scores` must wait. The transactions are now waiting on each other. Deadlock! |
+
+We can avoid deadlock by noting that writes from a provider DAG will necessarily
+_always have more up-to-date data_: so when a popularity refresh task encounters
+a locked row, we can tell it to simply skip that record using `SKIP LOCKED`
+(example code below).
+
+We have now avoided deadlock but we still have a problem with our example: the
+Flickr reingestion DAG must still wait for `refresh_flickr_scores` to release
+the lock on record B. Our upsert task will hang and likely time out, as the
+popularity refresh task will take a considerably long time to complete.
+
+To avoid this, we must break up the popularity refresh into smaller
+transactions, committing after every 10,000 rows are updated (this number can be
+tweaked as necessary). This ensures that a concurrently running provider DAG
+will only ever have to wait for a second or two before locked rows are released.
+
+The approach is easiest explained with some sample code:
+
+```pgsql
+-- The final product will be generalized by provider and media type.
+-- In a local test, this successfully updated ~1 million Flickr
+-- records in ~6 seconds, while the reingestion DAG was running.
+DO $$
+DECLARE
+  rows_per_loop int := 10000;
+  min_row int := 1;
+  max_row int;
+  max_timestamp timestamp;
+BEGIN
+  -- We will use this to select rows that have not been recently updated.
+  SELECT NOW() INTO max_timestamp;
+
+  -- A temporary table storing the identifiers of all records that need
+  -- to be updated, plus their row number. The row number will be used to
+  -- paginate the results more efficiently than using OFFSET.
+  CREATE TEMP TABLE rows_to_update AS
```

Comment:
Since multiple providers might be updated at once, does this table identifier need to be unique for each provider?

Reply:
Good point, I'll make sure to specify that!